### ISSUE TITLE
Fix Region escape check

### ISSUE THREAD

**Issue by JonathanStarup** [flix/flix#8781]
We currently check if `r` is escaping in `f` by saying that it does *not* if a) `r` is not syntactically present and b) `f[r -> True] =B= f[r -> Fales`.

Both of these are incorrect. If you have effect `not Crash` then the region has escaped (even though it isn't syntactically there). If you have `Univ - r` then you don't escape, even though `f[r -> True] != f[r -> False]`.

**Comment by JonathanStarup** [comment]
@magnus-madsen I think this is important but maybe we should still wait until sets are in?

**Comment by magnus-madsen** [comment]
The new system will use levels. I think that should solve the problem (?)

**Comment by JonathanStarup** [comment]
it still uses post checking for type escaping (from what ive discussed with matt)

**Comment by magnus-madsen** [comment]
I agree that the two checks are broken. I think that was already know. With levels it should be clear, but I don't know when the actual check is performed.

**Comment by JonathanStarup** [comment]
If the returned type of a region is `Array[Int32, r]` how does levels help?

**Comment by magnus-madsen** [comment]
> If the returned type of a region is `Array[Int32, r]` how does levels help?

We don't even have to check the return types. The level of `r` will simply be lower than where it was declared.

**Comment by magnus-madsen** [comment]
> So that means that `Univ - r` would be flagged as an escape? since its a syntactic unification trigger?

Yes, most definitely.

Can you come up with a real example where the effect `Univ - r` occurs outside the scope of region r?

**Comment by JonathanStarup** [comment]
- This is enough to break soundness/safety in a formalization so it seems nice to figure out the correct thing
- As you said, its syntactic, so levels would not say that `not Crash` is escaping? but it is. So its both imprecise and unsound.
- I don't know how much of a real example you want, but with `type alias Rev[a,b,ef] = a -> b \ ~ef` I can cheat your level checks

**Comment by mlutze** [comment]
This is fixed by the new rules preventing unification across region boundaries.
